packages/react-devtools-shared/src/backend/views/TraceUpdates/canvas.js

```diff
@@ -46,62 +44,132 @@ function drawWeb(nodeToData: Map<HostInstance, Data>) {
+  canvasFlow.height = window.innerHeight * dpr;
+  canvasFlow.style.width = `${window.innerWidth}px`;
+  canvasFlow.style.height = `${window.innerHeight}px`;
 
   const context = canvasFlow.getContext('2d');
-  context.clearRect(0, 0, canvasFlow.width, canvasFlow.height);
-  iterateNodes(nodeToData, (rect, color) => {
-    if (rect !== null) {
-      drawBorder(context, rect, color);
-    }
+  context.scale(dpr, dpr);
+
+  context.clearRect(0, 0, canvasFlow.width / dpr, canvasFlow.height / dpr);
+
+  const mergedNodes = groupAndSortNodes(nodeToData);
+
+  mergedNodes.forEach(group => {
+    drawGroupBorders(context, group);
+    drawGroupLabel(context, group);
+  });
+}
+
+type GroupItem = {
+  rect: Rect,
+  color: string,
+  displayName: string | null,
+  count: number,
+};
+
+export function groupAndSortNodes(
+  nodeToData: Map<HostInstance, Data>,
+): Array<Array<GroupItem>> {
+  const positionGroups: Map<string, Array<GroupItem>> = new Map();
+
+  iterateNodes(nodeToData, ({rect, color, displayName, count}) => {
+    if (!rect) return;
+    const key = `${rect.left},${rect.top}`;
+    if (!positionGroups.has(key)) positionGroups.set(key, []);
+    positionGroups.get(key)?.push({rect, color, displayName, count});
+  });
+
+  return Array.from(positionGroups.values()).sort((groupA, groupB) => {
+    const maxCountA = Math.max(...groupA.map(item => item.count));
+    const maxCountB = Math.max(...groupB.map(item => item.count));
+    return maxCountA - maxCountB;
+  });
+}
+
+function drawGroupBorders(
+  context: CanvasRenderingContext2D,
+  group: Array<GroupItem>,
+) {
+  group.forEach(({color, rect}) => {
+    context.beginPath();
+    context.strokeStyle = color;
+    context.rect(rect.left, rect.top, rect.width - 1, rect.height - 1);
+    context.stroke();
   });
 }
 
+function drawGroupLabel(
+  context: CanvasRenderingContext2D,
+  group: Array<GroupItem>,
+) {
+  const mergedName = group
+    .map(({displayName, count}) =>
+      displayName ? `${displayName}${count > 1 ? ` x${count}` : ''}` : '',
+    )
+    .filter(Boolean)
+    .join(', ');
+
+  if (mergedName) {
+    drawLabel(context, group[0].rect, mergedName, group[0].color);
+  }
+}
+
 export function draw(nodeToData: Map<HostInstance, Data>, agent: Agent): void {
   return isReactNativeEnvironment()
     ? drawNative(nodeToData, agent)
     : drawWeb(nodeToData);
 }
 
+type DataWithColorAndNode = {
+  ...Data,
+  color: string,
+  node: HostInstance,
+};
+
 function iterateNodes(
   nodeToData: Map<HostInstance, Data>,
-  execute: (rect: Rect | null, color: string, node: HostInstance) => void,
+  execute: (data: DataWithColorAndNode) => void,
 ) {
-  nodeToData.forEach(({count, rect}, node) => {
-    const colorIndex = Math.min(COLORS.length - 1, count - 1);
+  nodeToData.forEach((data, node) => {
+    const colorIndex = Math.min(COLORS.length - 1, data.count - 1);
     const color = COLORS[colorIndex];
-    execute(rect, color, node);
+    execute({...data, color, node});
```

Comment:
I can convert it back to attributes if needed. I went with an object here because it felt more type-safe and flexible, but let me know what you think.

Reply:
> I can convert it back to attributes if needed. I went with an object here because it felt more type-safe and flexible, but let me know what you think.

This might become expensive, if the whole page re-renders? For each HTMLElement you will create a temporary copy here just for iteration purposes.

Reply:
Sorry for the delay, I’ve been unwell. I replaced the spread operator with explicit object construction to avoid extra copies. Let me know if this solves the issue for you.